### ISSUE TITLE
Propose on_evaluate_end for Callback

### ISSUE THREAD

**Issue by joelthchao** [fchollet/keras#2555]
Might be a solution for #2548 and #2521 
Sometimes we may need to do something more than printing metrics after evaluation in every end of epoch. For example, visualization, testing and printing customize information without ruin progress bar, however, the callbacks execution order is almost fixed in the code:

``` python
# in training.py
callbacks = [cbks.BaseLogger()] + callbacks + [self.history]
if verbose:
    callbacks += [cbks.ProgbarLogger()]
callbacks = cbks.CallbackList(callbacks)
```

Define callback execution order may expose too much implementation details for user, which makes Keras too complicated. I think we can add `on_evaluate_end` for `Callback` and `CallbackList`.

``` python
# class Callback
def on_evaluate_end(self, epoch, logs={}):
    pass

# class CallbackList
def on_evaluate_end(self, epoch, logs={}):
    for callback in self.callbacks:
        callback.on_evaluate_end(epoch, logs)
```

and add the entry point to `_fit_loop`

``` python
# _fit_loop() in training.py 
callbacks.on_epoch_end(epoch, epoch_logs)
if do_validation:
    callbacks.on_evaluate_end(epoch, epoch_logs)
if callback_model.stop_training:
    break
```

Then users can have more flexibility on building their callbacks.
(Of course we can add `on_evaluate_begin` as well.)

**Comment by himanshuhkr** [comment]
If it can be implemented it would help more to analyse the network.

**Comment by Cerno-b** [comment]
Makes sense to me. If I'm not mistaken, this is also the cause why custom outputs like edit distance for CTC support is crapping up the output:

`Epoch 904/1000
37/38 [============================>.] - ETA: 0s - loss: 8.6462e-04
Out of 256 samples:  Mean edit distance: 1.203 Mean normalized edit distance: 0.131
38/38 [==============================] - 6s - loss: 8.6978e-04 - val_loss: 0.9349`
